modal/image.py

```diff
@@ -2013,8 +2013,12 @@ def env(self, vars: dict[str, str]) -> "_Image":
         """
 
         def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
-            commands = ["FROM base"] + [f"ENV {key}={shlex.quote(val)}" for (key, val) in vars.items()]
-            return DockerfileSpec(commands=commands, context_files={})
+            try:
+                env_commands = [f"ENV {key}={shlex.quote(val)}" for (key, val) in vars.items()]
+            except TypeError:
+                non_str_keys = [key for key, val in vars.items() if not isinstance(val, str)]
+                raise InvalidError(f"Environment variables must be strings. Invalid keys: {non_str_keys}")
```

Comment:
Is there a reason you're deferring this till after catching a `TypeError`? One advantage to doing it immediately (i.e. not inside the `build_dockerfile` closure) is that the error will get attributed to the relevant line of user code; raising from the closure produces a very long and basically uninterpretable traceback.

I guess the downside is that code outside of the closure will execute on container cold starts which is undesirable. Shouldn't be too costly in most cases .... but every millisecond counts ;)